flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/CheckpointBarrierUnaligner.java

```diff
@@ -0,0 +1,427 @@
+
+		threadSafeUnaligner = new ThreadSafeUnaligner(totalNumChannels,	checkNotNull(channelStateWriter), this);
+	}
+
+	@Override
+	public void releaseBlocksAndResetBarriers() {
+		// make sure no additional data is persisted
+		Arrays.fill(barrierConsumedChannels, true);
+		// the next barrier that comes must assume it is the first
+		numBarrierConsumed = 0;
+		threadSafeUnaligner.releaseReceivedBarriers();
+	}
+
+	/**
+	 * For unaligned checkpoint, it never blocks processing from the task aspect.
+	 *
+	 * <p>For PoC, we do not consider the possibility that the unaligned checkpoint would
+	 * not perform due to the max configured unaligned checkpoint size.
+	 */
+	@Override
+	public boolean isBlocked(int channelIndex) {
+		return false;
+	}
+
+	/**
+	 * We still need to trigger checkpoint while reading the first barrier from one channel, because this might happen
+	 * earlier than the previous async trigger via mailbox by netty thread. And the {@link AbstractInvokable} has the
+	 * deduplication logic to guarantee trigger checkpoint only once finally.
+	 *
+	 * <p>Note this is also suitable for the trigger case of local input channel.
+	 */
+	@Override
+	public boolean processBarrier(
+			CheckpointBarrier receivedBarrier,
+			int channelIndex,
+			long bufferedBytes) {
+		long barrierId = receivedBarrier.getId();
+		if (currentConsumedCheckpointId < barrierId) {
+			currentConsumedCheckpointId = barrierId;
+			numBarrierConsumed = 0;
+			Arrays.fill(barrierConsumedChannels, false);
+		}
+		if (currentConsumedCheckpointId == barrierId) {
+			barrierConsumedChannels[channelIndex] = true;
+			numBarrierConsumed++;
+		}
+		// processBarrier is called from task thread and can actually happen before notifyBarrierReceived on empty
+		// buffer queues
+		// to avoid replicating any logic, we simply call notifyBarrierReceived here as well
+		threadSafeUnaligner.notifyBarrierReceived(receivedBarrier, channelInfos[channelIndex]);
+		return false;
+	}
+
+	@Override
+	public boolean processCancellationBarrier(CancelCheckpointMarker cancelBarrier) throws Exception {
+		final long barrierId = cancelBarrier.getCheckpointId();
+
+		if (numBarrierConsumed > 0) {
+			// this is only true if some alignment is in progress and nothing was canceled
+
+			if (barrierId == currentConsumedCheckpointId) {
+				// cancel this alignment
+				if (LOG.isDebugEnabled()) {
+					LOG.debug("{}: Checkpoint {} canceled, aborting alignment.", taskName, barrierId);
+				}
+
+				releaseBlocksAndResetBarriers();
+				notifyAbortOnCancellationBarrier(barrierId);
+			}
+			else if (barrierId > currentConsumedCheckpointId) {
+				// we canceled the next which also cancels the current
+				LOG.warn("{}: Received cancellation barrier for checkpoint {} before completing current checkpoint {}. " +
+						"Skipping current checkpoint.",
+					taskName,
+					barrierId,
+					currentConsumedCheckpointId);
+
+				// this stops the current alignment
+				releaseBlocksAndResetBarriers();
```

Comment:
In `this.releaseBlocksAndResetBarriers` we call
`threadSafeUnaligner.releaseReceivedBarriers();`

But in `threadSafeUnaligner` condition `barrierId > currentConsumedCheckpointId` may not hold.
Should we add a check in `threadSafeUnaligner.releaseReceivedBarriers` similar to `setCurrentReceivedCheckpointId`?

nit: repeating method name for a nested class (discussed above)

Reply:
I added the checkpoint id as a parameter to make sure they talk about the same thing.

The current method name is `resetReceivedBarriers`. I think it's now quite precise but feel free to add if it's still confusing.